### DISCUSSION TITLE
Button OnClicked Extension

### DISCUSSION THREAD

**Discussion by Matthew-Collins** [CommunityToolkit/Maui.Markup#124] [New Feature Discussions]
I'm new to Maui and C# (VisualBasic developer 30 years) and struggled to understand how to wire up a click event... considered going back to XAML (found more support online) but eventually found the solution on one of your YouTube videos https://youtu.be/yS8XmAQi3YY?t=357 5:49, using the .Invoke(). 

My suggestion is to add a little extension to make this use beautiful (see .OnClicked). 

Example Usage:
```
new Button()
    .Text("Submit")
    //.Invoke(button => button.Clicked += OnSubmitClicked)
    .OnClicked(OnSubmitClicked)
```

```
private async void OnSubmitClicked(object sender, EventArgs e)
{
   //....
}
```

Extension Code:
```
public static TButton OnClicked<TButton>(this TButton button, EventHandler onClick) where TButton : Button
{
    button.Clicked += onClick;
    return button;
}  
```

**Comment by TheCodeTraveler** [comment]
I like it!

The concerns I have are inducing memory leaks, and unsubscribing from the event. 

The same functionality is currently allowed in XAML and doesn't seem to cause any complaints. Maybe we could implement it the same way that the .NET MAUI team wires up events subscribed in XAML?

```xml
<Button Clicked="HandleClicked" />
```

XAML also doesn't allow users to unsubscribe, so perhaps that shouldn't be a blocker.

**Reply by pictos** [reply]
I'm not against of having this extension, but I believe it's good bring more context and open the eyes to other scenarios
Depending on how devs use that extension method can cause memory leaks. 

On the Page it doesn't cause memory leak, because de lifecycle of all parts are the same, so you have something like

```csharp
class Mypage : ContentPage
{
     Button btn = new();

     public MyPage()
     {
          btn.Clicked += (s, e) => Console.WriteLine("Bla");
     }
}
```

So, here, `btn` is rooted to `MyPage`, so when the `Page` is no longer reachable the whole thing is cleaned up.

But if the dev uses this?
```csharp
class Mypage : ContentPage
{
     Button btn = new();

     public MyPage(SomeControl control)
     {
          control.Button.Clicked += OnClicked;
     }
     
     public void OnClicked(object sender, EventArgs e) { }
}

// Tested using the objectId from Rider
```

Now the life cycle is different and the page can leak. So in this case I would suggest to add this into the documentation in order to educate the dev on how this method should not be used.

**Comment by CliffAgius** [comment]
Closed as per the Community Stand-up discussion here: https://youtu.be/N9wMcBP4jtg?t=2889
